### ISSUE TITLE
Do not show "Enable location services" on every launch

### ISSUE THREAD

**Issue by biodranik** [organicmaps/organicmaps#1128]
Some users do not enable Location services at all and always see the "Please enable Location Services" annoying dialog on every app launch. The dialog appears anyway if the user presses the "my location" button in the app, so for better UX it makes sense to show dialog only once and store this fact in the settings, but reset this setting again if the location was successfully acquired (e.g. user enabled location services).

The main idea is to show this warning only once, if the user does not enable location services, and show it again if services were enabled, and then disabled again.

**Comment by rtsisyk** [comment]
> Some users do not enable Location services at all and always see the "Please enable Location Services" annoying dialog on every app launch

The app sets "LocationRequested" config flag after the first request from SplashActivity and doesn't try to request location anymore. However, location can be requested if user presses "Arrow" button or starts routing. I think that we can close this ticket.

**Comment by jplatte** [comment]
I would go one stop further even, I would prefer if OMaps didn't try to access my location un-prompted, regardless of whether the OS setting is on or not. It wastes energy when I don't need it, which is very often the case (for example: I want to look up the address of a location where I could stay on a trip, to see points of interest nearby). I'll open a new issue for that.

**Comment by biodranik** [comment]
The same issue happens again. 

Regression is confirmed with the latest fdroid build (didn't check with google build):
1. Disable location in the system. 
2. Run ОМ. See the dialog. 
3. Close OM. Kill OM. 
4. Run OM. Dialog is shown again. 

Dialog should be visible again after a long period of time (8 or more hours, there should be a constant in the code).

@rtsisyk

**Comment by biodranik** [comment]
> When location is off, OrganicMaps prompts me to turn on location when I open it. That's not a big deal; I dismiss it with "CLOSE".
 > But, in this new version, a few seconds or minutes later, I get a prompt like "Continue detecting your location? You may be in a tunnel". I think this shouldn't appear if my location is off. 😹

**Comment by rtsisyk** [comment]
> The same issue happens again.
> 
> Regression is confirmed with the latest fdroid build (didn't check with google build):
> 
>     1. Disable location in the system.
> 
>     2. Run ОМ. See the dialog.
> 
>     3. Close OM. Kill OM.
> 
>     4. Run OM. Dialog is shown again.

Is it the same issue as https://github.com/organicmaps/organicmaps/issues/6098? Can we close this ticket as a duplicate?

**Comment by biodranik** [comment]
No, this is a different one. The timer that was used before to avoid reminding users to enable location services (in 30 hours?) is broken now.

**Comment by rtsisyk** [comment]
> No, this is a different one. The timer that was used before to avoid reminding users to enable location services (in 30 hours?) is broken now.

I am not aware of the existence of such a 30-hours timer. I'm not entirely certain if it has ever been present in Organic Maps. No refactoring in the past 6 months altered that imaginary timer even if we assume that it existed. 

What can be reliably confirmed is that the app does not even attempt to enable location services on startup if the location was previously in the "disabled" state (i.e., the location icon was crossed out). The only issue here is that there is no straightforward way to enter that "disabled" mode, apart from pressing "STOP" in the "Continue detecting your location?" dialog.  And any press of the location button or attempt to build a route will re-enable location back. In other words, Organic Maps doesn't have any explicit reliable setting to disable location services. The crossed out icon exists, but it just confuses the people and causes annoying bugs like #5024. Probably it is time for the creators of "crossed out mode" to go back to the drawing board and re-design the broken feature.

**Comment by biodranik** [comment]
drape_frontend/my_position_controller.cpp
```
double const kMaxTimeInBackgroundSec = 60.0 * 60 * 30;  // 30 hours before starting detecting position again
```

**Comment by rtsisyk** [comment]
> drape_frontend/my_position_controller.cpp
> 
> ```
> double const kMaxTimeInBackgroundSec = 60.0 * 60 * 30;  // 30 hours before starting detecting position again
> ```

Nice to know that drape manipulates with this important settings implicitly using this muddy logic:

https://github.com/organicmaps/organicmaps/blob/cbafb7cc10ac62533f84b2fd891f3124af9d12ff/drape_frontend/my_position_controller.cpp#L154-L174

It explains the situation why NoFollowNoPosition is reset periodically, causing users who don't want to use location to see the location dialogs again every 30 hours. That means that the 'disabled' location mode with a crossed-out icon is just fundamentally flawed in its design.

--------------------

This particular bug happens because `g_framework->GetMyPositionMode()` returns PendingLocation instead of the last saved setting when it's called before the constructor of  MyPositionLocation:

1. Start the app first time

```
I app.organicmaps.location: MwmActivity.java:1837 autostartLocation(): Requesting ACCESS_FINE_LOCATION + ACCESS_FINE_LOCATION permissions
D app.organicmaps.location: LocationHelper.java:219 addListener(): listener: app.organicmaps.MwmActivity@ef62cf9 count was: 0
D app.organicmaps.location: MwmActivity.java:1697 onMyPositionModeChanged(): newMode = 0
W app.organicmaps.location: MwmActivity.java:1860 onLocationPermissionsResult(): Permission ACCESS_COARSE_LOCATION has been refused
W app.organicmaps.location: MwmActivity.java:1860 onLocationPermissionsResult(): Permission ACCESS_FINE_LOCATION has been refused
W app.organicmaps.location: MwmActivity.java:1869 onLocationPermissionsResult(): Permissions ACCESS_COARSE_LOCATION and ACCESS_FINE_LOCATION have been refused
D app.organicmaps.location: LocationHelper.java:309 stop(): Already stopped
D app.organicmaps.location: MwmActivity.java:1697 onMyPositionModeChanged(): newMode = 1
```

newMode = 1  == NoFollowNoPosition

2. Switch to any other app to make sure that all settings are gracefully saved to the config file.

3. Kill the app and start it again

```
D app.organicmaps.location: MwmActivity.java:1697 onMyPositionModeChanged(): newMode = 0
I OMcore  : drape_frontend/my_position_controller.cpp:177 MyPositionController(): init location mode 1 desired location mode 1
W app.organicmaps.location: MwmActivity.java:1861 onLocationPermissionsResult(): Permission ACCESS_COARSE_LOCATION has been refused
W app.organicmaps.location: MwmActivity.java:1861 onLocationPermissionsResult(): Permission ACCESS_FINE_LOCATION has been refused
W app.organicmaps.location: MwmActivity.java:1870 onLocationPermissionsResult(): Permissions ACCESS_COARSE_LOCATION and ACCESS_FINE_LOCATION have been refused
D app.organicmaps.location: LocationHelper.java:309 stop(): Already stopped
D app.organicmaps.location: MwmActivity.java:1697 onMyPositionModeChanged(): newMode = 1
```

mode == PendingLocation

MyPositionController() is called by CreateDrapeEngine() which is called some time later when Map view is created.

**Comment by rtsisyk** [comment]
Fixed by starting location after initialization of the drape.

**Comment by nach0** [comment]
Unfortunately this issue is back in the newest beta (version 2024.02.05-19-FDroid-beta). It was properly solved in November, and the current stable version (2023.12.20-4-FDroid) doesn't suffer from this problem.

**Comment by biodranik** [comment]
The issue is caused by recent iOS-related changes where "Continue detecting your location?" dialog was completely removed, and C++ core part was slightly modified to implement an option to turn location off by tapping on the "location crosshair" button: https://github.com/organicmaps/organicmaps/pull/7188

CC @rtsisyk @Jean-BaptisteC
